pkcs8/src/private_key_info.rs

```diff
@@ -250,45 +290,127 @@ impl<'a> fmt::Debug for PrivateKeyInfo<'a> {
+{
     type Error = Error;
 
-    fn try_from(private_key: PrivateKeyInfo<'_>) -> Result<SecretDocument> {
+    fn try_from(private_key: PrivateKeyInfo<Params, Key, PubKey>) -> Result<SecretDocument> {
         SecretDocument::try_from(&private_key)
     }
 }
 
 #[cfg(feature = "alloc")]
-impl TryFrom<&PrivateKeyInfo<'_>> for SecretDocument {
+impl<'a, Params, Key, PubKey> TryFrom<&PrivateKeyInfo<Params, Key, PubKey>> for SecretDocument
+where
+    Params: der::Choice<'a, Error = der::Error> + Encode,
+    Key: DecodeValue<'a, Error = der::Error> + FixedTag + 'a,
+    Key: EncodeValue,
+    PubKey: DecodeValue<'a, Error = der::Error> + FixedTag + 'a,
+    PubKey: BitStringLike,
+{
     type Error = Error;
 
-    fn try_from(private_key: &PrivateKeyInfo<'_>) -> Result<SecretDocument> {
+    fn try_from(private_key: &PrivateKeyInfo<Params, Key, PubKey>) -> Result<SecretDocument> {
         Ok(Self::encode_msg(private_key)?)
     }
 }
 
 #[cfg(feature = "pem")]
-impl PemLabel for PrivateKeyInfo<'_> {
+impl<Params, Key, PubKey> PemLabel for PrivateKeyInfo<Params, Key, PubKey> {
     const PEM_LABEL: &'static str = "PRIVATE KEY";
 }
 
 #[cfg(feature = "subtle")]
-impl<'a> ConstantTimeEq for PrivateKeyInfo<'a> {
+impl<Params, Key, PubKey> ConstantTimeEq for PrivateKeyInfo<Params, Key, PubKey>
+where
+    Params: Eq,
+    Key: PartialEq + AsRef<[u8]>,
+    PubKey: PartialEq,
+{
     fn ct_eq(&self, other: &Self) -> Choice {
         // NOTE: public fields are not compared in constant time
         let public_fields_eq =
             self.algorithm == other.algorithm && self.public_key == other.public_key;
 
-        self.private_key.ct_eq(other.private_key) & Choice::from(public_fields_eq as u8)
+        self.private_key.as_ref().ct_eq(other.private_key.as_ref())
+            & Choice::from(public_fields_eq as u8)
     }
 }
 
 #[cfg(feature = "subtle")]
-impl<'a> Eq for PrivateKeyInfo<'a> {}
+impl<Params, Key, PubKey> Eq for PrivateKeyInfo<Params, Key, PubKey>
+where
+    Params: Eq,
+    Key: AsRef<[u8]> + Eq,
+    PubKey: Eq,
+{
+}
 
 #[cfg(feature = "subtle")]
-impl<'a> PartialEq for PrivateKeyInfo<'a> {
+impl<Params, Key, PubKey> PartialEq for PrivateKeyInfo<Params, Key, PubKey>
+where
+    Params: Eq,
+    Key: PartialEq + AsRef<[u8]>,
+    PubKey: PartialEq,
+{
     fn eq(&self, other: &Self) -> bool {
         self.ct_eq(other).into()
     }
 }
+
+/// [`PrivateKeyInfo`] with [`AnyRef`] algorithm parameters, and `&[u8]` key.
+pub type PrivateKeyInfoRef<'a> = PrivateKeyInfo<AnyRef<'a>, OctetStringRef<'a>, BitStringRef<'a>>;
+
+/// [`PrivateKeyInfo`] with [`Any`] algorithm parameters, and `Box<[u8]>` key.
+#[cfg(feature = "alloc")]
+pub type PrivateKeyInfoOwned = PrivateKeyInfo<Any, OctetString, BitString>;
+
+pub trait BitStringLike {
+    fn as_bit_string(&self) -> BitStringRef<'_>;
+}
+
+impl<'a> BitStringLike for BitStringRef<'a> {
+    fn as_bit_string(&self) -> BitStringRef<'_> {
+        BitStringRef::from(self)
+    }
+}
```

Comment:
I could not figure out a way to express this with the trait implemented by both `BitStringRef` and `BitString`.

Reply:
In the interest of getting this PR across the finish line I'll go ahead and merge this, but I think we should definitely find a better solution, and it should go in the `der` crate.

Something like `OwnedToRef` but impl'd for both the owned and reference types so it can be used interchangeably, and for any of the `*Ref` types and not just specialized to `BitString*`.